# 🏏 IPL Cricket Data Analysis
## Complete Exploratory Data Analysis


**Dataset:** IPL Complete Dataset (Kaggle)

**Tools:** Python, Pandas, NumPy, Matplotlib, Seaborn

### Project Goal
Analyze IPL cricket matches from 2008-2022
to find patterns in team performance, venue
effects, toss impact and scoring trends.

## Step 1: Import Libraries

Importing all required Python libraries:
- **Pandas** → data manipulation and analysis
- **NumPy** → numerical computations
- **Matplotlib** → basic plotting
- **Seaborn** → statistical visualizations

In [28]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Step 2: Load Dataset

Loading two CSV files:
- **matches.csv** → one row per IPL match
- **deliveries.csv** → one row per ball bowled

Both files placed in `data/raw/` folder
(raw data — never modified!)

In [71]:

df1=pd.read_csv('../data/raw/matches.csv',index_col=False)
df2=pd.read_csv('../data/raw/deliveries.csv',index_col=False)


## Step 3: First Look at Data

Checking basic information about both datasets:
- Shape (rows × columns)
- Column names
- Data types
- Sample rows

**Goal:** Understand what data we have
before any analysis!

In [55]:
print("Matches dataset columns:")
print(list(df1.columns))
print("Deliveries dataset columns:")
print(list(df2.columns))

print("Matches dataset shape:", df1.shape)
print("Deliveries dataset shape:", df2.shape)
print("Matches dataset data types:")
print(df1.dtypes)
print("Deliveries dataset data types:")
print(df2.dtypes)


Matches dataset columns:
['id', 'season', 'city', 'date', 'match_type', 'player_of_match', 'venue', 'team1', 'team2', 'toss_winner', 'toss_decision', 'winner', 'result', 'result_margin', 'target_runs', 'target_overs', 'super_over', 'method', 'umpire1', 'umpire2']
Deliveries dataset columns:
['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball', 'batter', 'bowler', 'non_striker', 'batsman_runs', 'extra_runs', 'total_runs', 'extras_type', 'is_wicket', 'player_dismissed', 'dismissal_kind', 'fielder']
Matches dataset shape: (1095, 20)
Deliveries dataset shape: (260920, 17)
Matches dataset data types:
id                   int64
season                 str
city                   str
date                   str
match_type             str
player_of_match        str
venue                  str
team1                  str
team2                  str
toss_winner            str
toss_decision          str
winner                 str
result                 str
result_margin      float64
ta

## Step 4: Data Quality Check

Checking for:
- **Missing values** → which columns? how many?
- **Duplicate rows** → any repeated data?
- **Data types** → correct types for each column?

**Goal:** Identify problems BEFORE analysis
so results are reliable!

In [60]:
print("Matches data set null checks",df1.isnull().sum())
print("============"*50)
print("Deliveries data set null checks",df2.isnull().sum())
print("============"*50)

print("Matches dataset duplicate values:")
print(df1.duplicated().any())

print("Count of duplicate rows in Matches dataset:")
# Count EXACTLY how many duplicates
print(df1.duplicated().sum())
print("============"*50)

print("Deliveries dataset duplicate values:")
print(df2.duplicated().any())
print("============"*50)

print("Count of duplicate rows in Deliveries dataset:")
# Count EXACTLY how many duplicates
print(df2.duplicated().sum())
print("============"*50)

print("Matches dataset info:")
print(df1.info())
print("============"*50)

print("Deliveries dataset info:")
print(df2.info())
print("============"*50)

print("Matches dataset top 5 rows:")
print(df1.head().to_string(index=False))
print("============"*50)
print("Matches dataset bottom 5 rows:")
print(df1.tail().to_string(index=False))
print("============"*50)



print("Deliveries dataset top 5 rows:")
print(df2.head().to_string(index=False))
print("============"*50)
print("Deliveries dataset bottom 5 rows:")
print(df2.tail().to_string(index=False))


Matches data set null checks id                    0
season                0
city                 51
date                  0
match_type            0
player_of_match       5
venue                 0
team1                 0
team2                 0
toss_winner           0
toss_decision         0
winner                5
result                0
result_margin        19
target_runs           3
target_overs          3
super_over            0
method             1074
umpire1               0
umpire2               0
dtype: int64
Deliveries data set null checks match_id                 0
inning                   0
batting_team             0
bowling_team             0
over                     0
ball                     0
batter                   0
bowler                   0
non_striker              0
batsman_runs             0
extra_runs               0
total_runs               0
extras_type         246795
is_wicket                0
player_dismissed    247970
dismissal_kind      247970
fielder       

## Step 4B: Fix Missing Values — Matches Dataset

Based on our missing value analysis, decisions made:

| Column | Missing | % | Decision | Fill Value |
|--------|---------|---|----------|------------|
| city | 51 | 4.7% | Fill | 'Unknown' |
| player_of_match | 5 | 0.5% | Fill | 'Unknown' |
| winner | 5 | 0.5% | Fill | 'No Result' |
| result_margin | 19 | 1.7% | Fill | 0 |
| target_runs | 3 | 0.3% | Fill | 0 |
| target_overs | 3 | 0.3% | Fill | 0 |
| method | 1074 | 98% | **DROP** | Too many missing! |

**Rule applied:**
- Missing < 50% → Fill with appropriate value
- Missing > 50% → Drop the column

In [72]:
print("============"*50)

print("Before handling missing values in Matches dataset...")
print(df1.shape)
print(df1.isnull().sum())
df1['city']= df1['city'].fillna('Unknown')
df1['player_of_match']=df1['player_of_match'].fillna('Unknown')
df1['winner']=df1['winner'].fillna('No Result')
df1['result_margin']=df1['result_margin'].fillna(0)
df1['target_runs']=df1['target_runs'].fillna(0)
df1['target_overs']=df1['target_overs'].fillna(0)
df1 = df1.drop(columns=['method'])
print("After handling missing values, Matches dataset info:")
print("============"*50)
print(df1.shape)
print(df1.isnull().sum())

Before handling missing values in Matches dataset...
(1095, 20)
id                    0
season                0
city                 51
date                  0
match_type            0
player_of_match       5
venue                 0
team1                 0
team2                 0
toss_winner           0
toss_decision         0
winner                5
result                0
result_margin        19
target_runs           3
target_overs          3
super_over            0
method             1074
umpire1               0
umpire2               0
dtype: int64
After handling missing values, Matches dataset info:
(1095, 19)
id                 0
season             0
city               0
date               0
match_type         0
player_of_match    0
venue              0
team1              0
team2              0
toss_winner        0
toss_decision      0
winner             0
result             0
result_margin      0
target_runs        0
target_overs       0
super_over         0
umpire1            0
u

## Step 4C: Fix Missing Values — Deliveries Dataset

| Column | Missing | % | Decision | Reason |
|--------|---------|---|----------|--------|
| extras_type | 246,795 | 94.6% | **DROP** | Too many missing |
| fielder | 251,566 | 96.4% | **DROP** | Too many missing |
| player_dismissed | 247,970 | 95.1% | Fill 'not_out' | NaN = batsman survived! |
| dismissal_kind | 247,970 | 95.1% | Fill 'not_out' | NaN = not dismissed! |

**Special Rule Applied:**
> Missing > 50% normally means DROP
> BUT player_dismissed and dismissal_kind are SPECIAL!
> NaN here means batsman was NOT OUT that ball
> This is MEANINGFUL information — not truly missing!
> So we KEEP and fill with 'not_out' label

In [74]:
print("============"*50)
print("Before handling missing values in Deliveries dataset...")
print(df2.shape)
print(df2.isnull().sum())


df2=df2.drop(columns=['fielder'])
df2=df2.drop(columns=['extras_type'])
df2['player_dismissed']=df2['player_dismissed'].fillna('not out')
df2['dismissal_kind']=df2['dismissal_kind'].fillna('not out')
print("After handling missing values, Deliveries dataset info:")
print("============"*50)
print(df2.shape)
print(df2.isnull().sum())

Before handling missing values in Deliveries dataset...
(260920, 17)
match_id                 0
inning                   0
batting_team             0
bowling_team             0
over                     0
ball                     0
batter                   0
bowler                   0
non_striker              0
batsman_runs             0
extra_runs               0
total_runs               0
extras_type         246795
is_wicket                0
player_dismissed    247970
dismissal_kind      247970
fielder             251566
dtype: int64
After handling missing values, Deliveries dataset info:
(260920, 15)
match_id            0
inning              0
batting_team        0
bowling_team        0
over                0
ball                0
batter              0
bowler              0
non_striker         0
batsman_runs        0
extra_runs          0
total_runs          0
is_wicket           0
player_dismissed    0
dismissal_kind      0
dtype: int64


## Step 5: Save Cleaned Data

Saving both cleaned datasets to `data/processed/` folder.

**Why save separately from raw?**
- Raw data stays untouched always ✅
- Processed data ready for EDA ✅
- No need to repeat cleaning every time ✅
- Industry standard practice ✅

In [78]:
import os

df1.to_csv('../data/processed/matches_clean.csv',
           index=False)
df2.to_csv('../data/processed/deliveries_clean.csv',
           index=False)

print("✅ Both files saved!")

✅ Both files saved!


## ✅ Data Loading Complete — Summary

| Dataset | Original Shape | Clean Shape | Changes Made |
|---------|---------------|-------------|--------------|
| Matches | (1095, 20) | (1095, 19) | 6 filled, 1 dropped |
| Deliveries | (260920, 17) | (260920, 15) | 2 dropped, 2 filled |

### Key Findings From Data Quality Check:
1. Matches span **2007/08 to 2024** — 17 IPL seasons
2. **method column dropped** — 98% missing, not useful
3. **No duplicates** found in either dataset ✅
4. **Deliveries NaN** = meaningful (batsman not out)
5. Total balls bowled across all IPL = **260,920**

**Next Notebook:** `02_eda.ipynb` → Exploratory Analysis